File: examples/image_viewer/src/image_viewer.cpp
#include <atomic>
#include <csignal>
#include <iostream>
#include <nodar/zmq/image.hpp>
#include <nodar/zmq/opencv_utils.hpp>
#include <nodar/zmq/topic_ports.hpp>
#include <opencv2/highgui.hpp>
#include <unordered_map>
#include <zmq.hpp>

std::atomic_bool running{true};

void signalHandler(int signum) {
    std::cerr << "SIGINT or SIGTERM received." << std::endl;
    running = false;
}

const std::unordered_map<int, const char *> types = {
    {CV_8UC1, "CV_8UC1"},  //
    {CV_8UC3, "CV_8UC3"},  //
    {CV_8SC1, "CV_8SC1"},  //
    {CV_8SC3, "CV_8SC3"},  //
    {CV_16UC1, "CV_16UC1"},  //
    {CV_16UC3, "CV_16UC3"},  //
    {CV_16SC1, "CV_16SC1"},  //
    {CV_16SC3, "CV_16SC3"},  //
};

void printDetails(const cv::Mat &mat) {
    std::cout << mat.rows << ", " << mat.cols << ", " << mat.channels() << ", " << types.find(mat.type())->second
              << std::endl;
}

class ZMQImageViewer {
public:
    uint64_t last_frame_id = 0;

    ZMQImageViewer(const std::string &endpoint) : context(1), socket(context, ZMQ_SUB), window_name(endpoint) {
        const int hwm = 1;  // set maximum queue length to 1 message
        socket.set(zmq::sockopt::rcvhwm, hwm);
        socket.set(zmq::sockopt::subscribe, "");
        socket.connect(endpoint);
        std::cout << "Subscribing to " << endpoint << std::endl;
        cv::namedWindow(window_name, cv::WINDOW_NORMAL);
    }

    void loopOnce() {
        zmq::message_t msg;
        const auto received_bytes = socket.recv(msg, zmq::recv_flags::none);
        const nodar::zmq::StampedImage stamped_image(static_cast<uint8_t *>(msg.data()));
        auto img = nodar::zmq::cvMatFromStampedImage(stamped_image);
        if (img.type() == CV_16SC1) {
            // Highgui produces a strange-looking output for signed 16-bit images. Convert to unsigned
            img.convertTo(img, CV_16UC1);
        }
        if (img.empty()) {
            return;
        }
        const auto &frame_id = stamped_image.frame_id;
        if (last_frame_id != 0 and frame_id != last_frame_id + 1) {
            std::cerr << (frame_id - last_frame_id - 1) << " frames dropped. Current frame ID : " << frame_id
                      << ", last frame ID: " << last_frame_id << std::endl;
        }
        last_frame_id = frame_id;
        std::cout << "\rFrame # " << frame_id << std::flush;

        // Downsize the image before viewing
        cv::resizeWindow(window_name, {640, 480});
        cv::imshow(window_name, img);
        cv::waitKey(1);
        // You can try checking if the window is still visible, and stop if it is not.
        // However, that OpenCV function appears buggy on many systems.
        // If it is disabled, then you will have to CTRL+C in the terminal to kill it
        static constexpr bool stop_on_close = false;
        if (stop_on_close and cv::getWindowProperty(window_name, cv::WND_PROP_VISIBLE) < 1) {
            std::cout << "Stopping..." << std::endl;
            running = false;
        }
    }

private:
    zmq::context_t context;
    zmq::socket_t socket;
    std::string window_name;
};

constexpr auto DEFAULT_IP = "127.0.0.1";
constexpr auto DEFAULT_PORT = "9800";
constexpr auto DEFAULT_TOPIC = nodar::zmq::IMAGE_TOPICS[0];

void printUsage() {
    std::cout << "You should specify the Orin's IP address as well as \n"
                 "the port of the message that you want to listen to like this:\n\n"
                 "     ./image_viewer orin_ip port\n\n"
                 "e.g. ./image_viewer 192.168.1.9 9800\n\n"
                 "Alternatively, you can specify one of the image topic names provided in topic_ports.hpp of zmq_msgs:"
                 "e.g. ./image_viewer 192.168.1.9 nodar/right/image_raw\n\n"
                 "In the meantime, we are going to assume that you are running this on the Orin itself,\n"
                 "and that you want the images on port 9800, that is, we assume that you specified\n\n"
                 "     ./image_viewer 127.0.0.1 9800\n\n"
              << "\n\nNote that the list of topic/port mappings is in topic_ports.hpp header in the zmq_msgs target."
              << "\n----------------------------------------" << std::endl;
}

int main(int argc, char *argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);
    if (argc < 3) {
        printUsage();
    }
    const auto ip = argc > 1 ? argv[1] : DEFAULT_IP;

    // If no second argument was provided, then assume the default topic.
    // Try to parse the second argument to see if you provided a port number.
    nodar::zmq::Topic topic;
    if (argc < 3) {
        topic = DEFAULT_TOPIC;
    } else {
        // See if we can parse the second argument as a port number
        bool invalid_topic = true;
        size_t port = 0;
        std::istringstream iss(argv[2]);
        if (iss >> port) {
            for (const auto &image_topic : nodar::zmq::IMAGE_TOPICS) {
                if (port == image_topic.port) {
                    topic = image_topic;
                    invalid_topic = false;
                }
            }
            if (invalid_topic) {
                std::cerr << "It seems like you specified a port number " << port
                          << " that does not correspond to a port on which images are being published." << std::endl;
                return EXIT_FAILURE;
            }
        } else {
            // It seems like you specified a topic name. Let's see if it is a valid image topic
            const std::string topic_name = argv[2];
            for (const auto &image_topic : nodar::zmq::IMAGE_TOPICS) {
                if (topic_name == image_topic.name) {
                    topic = image_topic;
                    invalid_topic = false;
                }
            }
            if (invalid_topic) {
                std::cerr << "It seems like you specified a topic name " << topic_name
                          << " that does not correspond to a topic on which images are being published." << std::endl;
                return EXIT_FAILURE;
            }
        }
    }
    const auto endpoint = std::string("tcp://") + ip + ":" + std::to_string(topic.port);
    ZMQImageViewer subscriber(endpoint);
    while (running) {
        subscriber.loopOnce();
    }
}

File: examples/python/image_viewer/image_viewer/image_viewer.py
import os
import sys

import cv2
import numpy as np
import zmq

try:
    from zmq_msgs.image import StampedImage
    from zmq_msgs.topic_ports import IMAGE_TOPICS
except ImportError:
    sys.path.append(
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../zmq_msgs/python"))
    )
    from zmq_msgs.image import StampedImage
    from zmq_msgs.topic_ports import IMAGE_TOPICS

# Mapping of OpenCV types to strings
types = {
    0: "CV_8UC1",
    16: "CV_8UC3",
    1: "CV_8SC1",
    17: "CV_8SC3",
    2: "CV_16UC1",
    18: "CV_16UC3",
    3: "CV_16SC1",
    19: "CV_16SC3",
}


class ZMQImageViewer:
    def __init__(self, endpoint: str):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.SUB)
        self.socket.setsockopt(zmq.RCVHWM, 1)
        self.socket.setsockopt_string(zmq.SUBSCRIBE, "")
        self.socket.connect(endpoint)
        self.window_name = endpoint
        self.last_frame_id = 0
        print(f"Subscribing to {endpoint}")
        cv2.namedWindow(self.window_name, cv2.WINDOW_NORMAL)

    def loop_once(self):
        msg = self.socket.recv()
        stamped_image = StampedImage()
        stamped_image.read(msg)
        img = stamped_image.img
        if img.dtype == np.int16:
            # Highgui looks strange with signed 16-bit images. Convert to unsigned
            img = img.astype(np.uint16)
        if img is None or img.size == 0:
            print("img is None or img.size == 0")
            return True

        frame_id = stamped_image.frame_id
        if self.last_frame_id != 0 and frame_id != self.last_frame_id + 1:
            print(
                f"{frame_id - self.last_frame_id - 1} frames dropped. "
                f"Current frame ID: {frame_id}, last frame ID: {self.last_frame_id}"
            )
        self.last_frame_id = frame_id
        print(
            f"\rFrame # {frame_id}, img.shape = {img.shape}, img.dtype = {img.dtype}",
            end="",
            flush=True,
        )

        cv2.resizeWindow(self.window_name, 640, 480)
        cv2.imshow(self.window_name, img)
        cv2.waitKey(1)

        stop_on_close = False
        if stop_on_close and not cv2.getWindowProperty(self.window_name, cv2.WND_PROP_VISIBLE):
            print("Stopping...")
            return False
        return True


def print_usage(default_ip, default_port):
    print(
        "You should specify the IP address of the ZMQ source (the device running Hammerhead), \n"
        " as well as the port number of the message that you want to listen to:\n\n"
        "     python image_viewer.py hammerhead_ip port\n\n"
        "e.g. python image_viewer.py 192.168.1.9 9800\n\n"
        "Alternatively, you can specify one of the image topic names in topic_ports.py of zmq_msgs...\n\n"
        "e.g. python image_viewer.py 192.168.1.9 nodar/right/image_raw\n\n"
        "If unspecified, we assume you are running this on the device running Hammerhead,\n"
        "along with the defaults\n\n"
        f"     python image_viewer.py {default_ip} {default_port}\n\n"
        "Note that the list of topic/port mappings is in topic_ports.py in the zmq_msgs target.\n"
        "----------------------------------------"
    )


def main():
    default_ip = "127.0.0.1"
    default_topic = IMAGE_TOPICS[0]
    default_port = default_topic.port
    if len(sys.argv) < 3:
        print_usage(default_ip, default_port)

    ip = sys.argv[1] if len(sys.argv) > 1 else default_ip

    topic = default_topic
    if len(sys.argv) >= 3:
        try:
            port = int(sys.argv[2])
            for image_topic in IMAGE_TOPICS:
                if port == image_topic.port:
                    topic = image_topic
                    break
            else:
                print(
                    f"It seems like you specified a port number {port} "
                    "that does not correspond to a port on which images are being published."
                )
                return
        except ValueError:
            topic_name = sys.argv[2]
            for image_topic in IMAGE_TOPICS:
                if topic_name == image_topic.name:
                    topic = image_topic
                    break
            else:
                print(
                    f"It seems like you specified a topic name {topic_name} "
                    "that does not correspond to a topic on which images are being published."
                )
                return

    endpoint = f"tcp://{ip}:{topic.port}"
    subscriber = ZMQImageViewer(endpoint)
    try:
        while True:
            subscriber.loop_once()
    except KeyboardInterrupt:
        print("\nExiting...")


if __name__ == "__main__":
    main()
